source/io/Stream.cpp

```diff
@@ -13,132 +13,172 @@
+            };
 
-    return AWS_OP_SUCCESS;
-}
+            InputStream::InputStream(Aws::Crt::Allocator *allocator)
+            {
+                m_allocator = allocator;
+                m_underlying_stream =
+                    static_cast<aws_input_stream *>(aws_mem_calloc(m_allocator, 1, sizeof(aws_input_stream)));
 
-static void s_aws_input_stream_cpp_destroy(struct aws_input_stream *stream)
-{
-    aws_input_stream_cpp_impl *impl = static_cast<aws_input_stream_cpp_impl *>(stream->impl);
-    impl->stream = nullptr;
-    aws_mem_release(stream->allocator, stream);
-}
+                if (!m_underlying_stream)
+                {
+                    m_good = false;
+                    return;
+                }
 
-static struct aws_input_stream_vtable s_aws_input_stream_cpp_vtable = {s_aws_input_stream_cpp_seek,
-                                                                       s_aws_input_stream_cpp_read,
-                                                                       s_aws_input_stream_cpp_get_status,
-                                                                       s_aws_input_stream_cpp_get_length,
-                                                                       s_aws_input_stream_cpp_destroy};
+                m_good = true;
+                m_underlying_stream->impl = this;
+                m_underlying_stream->allocator = m_allocator;
+                m_underlying_stream->vtable = &s_vtable;
+            }
 
-namespace Aws
-{
-    namespace Crt
-    {
-        namespace Io
-        {
-            struct aws_input_stream *AwsInputStreamNewCpp(
-                const std::shared_ptr<Aws::Crt::Io::IStream> &stream,
+            StdIOStreamInputStream::StdIOStreamInputStream(
+                std::shared_ptr<Aws::Crt::Io::IStream> stream,
                 Aws::Crt::Allocator *allocator) noexcept
+                : InputStream(allocator), m_stream(std::move(stream))
             {
-                struct aws_input_stream *input_stream = NULL;
-                struct aws_input_stream_cpp_impl *impl = NULL;
-
-                aws_mem_acquire_many(
-                    allocator,
-                    2,
-                    &input_stream,
-                    sizeof(struct aws_input_stream),
-                    &impl,
-                    sizeof(struct aws_input_stream_cpp_impl));
-
-                if (!input_stream)
+            }
+
+            bool StdIOStreamInputStream::Read(ByteBuf &buffer) noexcept
+            {
+                m_stream->read(reinterpret_cast<char *>(buffer.buffer + buffer.len), buffer.capacity - buffer.len);
+                buffer.len += static_cast<size_t>(m_stream->gcount());
+
+                return true;
+            }
+
+            StreamStatus StdIOStreamInputStream::GetStatus() const noexcept
+            {
+                StreamStatus status;
+                status.is_end_of_stream = m_stream->eof();
+                status.is_valid = m_stream->good();
```

Comment:
Not 100% sure, but shouldn't you use `m_stream->bool()` instead of `m_stream->good()`?

Check out the graph at the bottom of [this page](https://en.cppreference.com/w/cpp/io/ios_base/iostate). The stream is no longer "good" once you get to eos, but is still "bool() == true".

PS: I am forever going to reference that graph when I hear we should handle errors like how std::iostream does it.